lib/Dialect/TTMetal/Transforms/Passes.cpp

```diff
@@ -16,44 +16,324 @@
 #include "ttmlir/Dialect/TTIR/IR/TTIR.h"
 #include "ttmlir/Dialect/TTIR/IR/TTIROps.h"
 #include "ttmlir/Dialect/TTIR/Transforms/Passes.h"
+#include "llvm/ADT/MapVector.h"
 
+#include "ttmlir/Dialect/TT/Utils/PhysicalCoreCoord.h"
 #include "ttmlir/Dialect/TTKernel/IR/TTKernel.h"
 #include "ttmlir/Dialect/TTKernel/IR/TTKernelOps.h"
 #include "ttmlir/Dialect/TTKernel/IR/TTKernelOpsTypes.h"
 #include "ttmlir/Dialect/TTMetal/IR/TTMetalOpsTypes.h"
+#include "ttmlir/Utils.h"
 
 namespace mlir::tt::ttmetal {
 
 #define GEN_PASS_DEF_CONVERTTTIRTOTTMETAL
 #include "ttmlir/Dialect/TTMetal/Transforms/Passes.h.inc"
 
+// This routine walks the SSA value chain to find the address of the value.
+// It runs into and gets the address from one of the following:
+//   - ttir::AllocOp: The address is the address of the allocation is embedded
+//   in the op attributes.
+//   - func::FuncOp Argument: The address is taken from ArgumentAllocationAttr.
+//
+// This routine is used to lookup the address of tensors for address generation
+// and for CB creation.
+static uint64_t lookupAddress(Value value) {
+  auto blockArg = mlir::dyn_cast<BlockArgument>(value);
+  if (blockArg) {
+    auto funcOp = blockArg.getOwner()->getParentOp();
+    if (mlir::isa<func::FuncOp>(funcOp)) {
+      auto argAlloc = mlir::cast<ArgumentAllocationAttr>(
+          mlir::cast<ArrayAttr>(funcOp->getDiscardableAttr(
+              ArgumentAllocationAttr::name))[blockArg.getArgNumber()]);
+      return argAlloc.getAddress();
+    }
+  }
+
+  auto *op = value.getDefiningOp();
+  if (!op) {
+    return 0;
+  }
+
+  while (isa<DestinationStyleOpInterface>(op)) {
+    assert(op->getResults().size() == 1);
+    auto dps = cast<DestinationStyleOpInterface>(op);
+    assert(dps.getNumDpsInits() == 1);
+    auto *opOperand = dps.getDpsInitOperand(0);
+    value = opOperand->get();
+    op = value.getDefiningOp();
+  }
+
+  auto allocOp = dyn_cast<ttir::AllocOp>(op);
+  if (!allocOp) {
+    return 0;
+  }
+  return allocOp.getAddress();
+}
+
 class TTIRToTTMetalLayoutRewriter : public OpRewritePattern<ttir::ToLayoutOp> {
 public:
   using OpRewritePattern<ttir::ToLayoutOp>::OpRewritePattern;
 
+  struct NocRead {
+    PhysicalCoreCoord srcCoord;
+    std::int64_t srcOffset = 0;
+    std::int64_t dstOffset = 0;
+    std::int64_t size = 0;
+
+    NocRead(PhysicalCoreCoord srcCoord, std::int64_t srcOffset,
+            std::int64_t dstOffset, std::int64_t size)
+        : srcCoord(srcCoord), srcOffset(srcOffset), dstOffset(dstOffset),
+          size(size) {}
+
+    bool isContiguous(PhysicalCoreCoord nextCoord, std::int64_t nextSrcOffset,
+                      std::int64_t nextDstOffset) const {
+      return (nextCoord == srcCoord) && (nextSrcOffset == srcOffset + size) &&
+             (nextDstOffset == dstOffset + size);
+    }
+  };
+
+  llvm::MapVector<PhysicalCoreCoord, mlir::SmallVector<NocRead>>
+  calculateDataMovement(ArrayRef<std::int64_t> tensorShape,
```

Comment:
Let's just add a comment about this function, as it's not obvious what it does on the first glance.